lib/std/os.zig

```diff
@@ -348,18 +348,154 @@ pub fn fchmod(fd: fd_t, mode: mode_t) FChmodError!void {
     if (!std.fs.has_executable_bit) @compileError("fchmodat unsupported by target OS");
 
     const path_c = try toPosixPath(path);
 
+    // No special handling for linux is needed if we can use the libc fallback
+    // or `flags` is empty.
+    if (builtin.os.tag != .linux or builtin.link_libc or flags == 0) {
+        while (true) {
+            const res = system.fchmodat(dirfd, &path_c, mode, flags);
+            switch (system.getErrno(res)) {
+                .SUCCESS => return,
+                .INTR => continue,
+                .BADF => unreachable,
+                .FAULT => unreachable,
+                .INVAL => unreachable,
+                .ACCES => return error.AccessDenied,
+                .IO => return error.InputOutput,
+                .LOOP => return error.SymLinkLoop,
+                .MFILE => return error.ProcessFdQuotaExceeded,
+                .NAMETOOLONG => return error.NameTooLong,
+                .NFILE => return error.SystemFdQuotaExceeded,
+                .NOENT => return error.FileNotFound,
+                .NOTDIR => return error.FileNotFound,
+                .NOMEM => return error.SystemResources,
+                .OPNOTSUPP => return error.OperationNotSupported,
+                .PERM => return error.AccessDenied,
+                .ROFS => return error.ReadOnlyFileSystem,
+                else => |err| return unexpectedErrno(err),
+            }
+        }
+    }
+
+    if ((comptime builtin.os.isAtLeast(.linux, .{ .major = 6, .minor = 6, .patch = 0 }) orelse false) and
+        has_fchmodat2_syscall.load(.Monotonic))
+    {
+        while (true) {
+            const res = system.fchmodat2(dirfd, &path_c, mode, flags);
+            switch (system.getErrno(res)) {
+                .SUCCESS => return,
+                .INTR => continue,
+                .BADF => unreachable,
+                .FAULT => unreachable,
+                .INVAL => unreachable,
+                .ACCES => return error.AccessDenied,
+                .IO => return error.InputOutput,
+                .LOOP => return error.SymLinkLoop,
+                .NOENT => return error.FileNotFound,
+                .NOMEM => return error.SystemResources,
+                .NOTDIR => return error.FileNotFound,
+                .OPNOTSUPP => return error.OperationNotSupported,
+                .PERM => return error.AccessDenied,
+                .ROFS => return error.ReadOnlyFileSystem,
+
+                .NOSYS => { // Use fallback.
+                    has_fchmodat2_syscall.store(false, .Monotonic);
+                    break;
+                },
+                else => |err| return unexpectedErrno(err),
+            }
+        }
+    }
+
+    // Fallback to changing permissions using procfs:
+    //
+    // 1. Open `path` as an `O.PATH` descriptor.
+    // 2. Stat the fd and check if it isn't a symbolic link.
+    // 3. Generate the procfs reference to the fd via `/proc/self/fd/{fd}`.
+    // 4. Pass the procfs path to `chmod` with the `mode`.
```

Comment:
Maybe factor this implementation out into a `fn fchmodat_fallback(..)` function?